src/cropper.sol

```diff
@@ -0,0 +1,478 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published
+// by the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+import "./crop.sol";
```

Comment:
> When reviewing this I have come across a pretty critical assumption that is pretty easy to violate. We are assuming `gems <= stake[usr]` at all times, but this can be pretty easily circumvented by having a user `flux` their gems without calling `tack`. I've added this problematic case as a failing test in the subsequent commit. Even if we remove the `tack()` calls from the liquidations we still have this problem where people could move gems around in a quite complex pattern and we leave it up to the keepers to tack these around from arbitrary addresses.
> 
> I think we need to do 2 things to solve this issue.
> 
> 1. Liquidations should not do anything fancy with the rewards. I think upon `kick` being called any outstanding rewards should just be sent to the vault owner (or maybe we just do nothing and the rewards are lost).
> 2. `flee()` and `exit()` need to allow for the case where `tack()` was not called and you just get the collateral back without rewards.

Yeah good catch. My thoughts:

1. Sending any rewards to the Vault owner on `kick` is fair--the rewards don't collateralize their position, so there's no real reason someone should lose those rewards if liquidated. The difficulty is the CDPManager thing Gonzalo brought up--we can really only send rewards back to the urn address, i.e. the UrnHandler, and the rewards would be lost anyway. The only fully satisfactory solution I see is unfortunately quite complex: the adapter itself should be something like the CdpManager, provisioning internally some address to control the Vault, and only exposing a subset of Vat UIs to end users (so e.g. external users could only `flux` via a function on the adapter that would enforce appropriate `tack` usage).

2. For `exit` and `flee`...need to think carefully here that there isn't e.g. some griefing vector. But if we went with the design above (adapter becomes a middleware that mediates access to the core APIs), the problem goes away here too.

Basically I now think the way we should have done it in the first place is what I describe in (1), but we're pretty far down this other path so there's an argument we should accept the suboptimal thing to get this out the door. I might spin off another branch to try out that idea and see just how complex it is.